Add kotlin observer pattern #5

Dhuds1 opened the pull request:
Hello community! I created a observer pattern in the kotlin language using an interface implementation. Please let me know if this chunk of code is up to standards, and if you're in need of any edits!
P.S I forgot to tag the issue #5
P.S.S There is no commit history because I was having issues working with the repository inside IntelliJ and it would only compile inside a fresh project.

Evelynn8Bit committed:
Create observer pattern

Evelynn8Bit committed:
Fix file structure

ash-teach commented:
@Dhuds1 - I just realized I must have neglected to ask you to include any references you used in developing this code. Can you update to include references in comments at the top of the file?

Could you also please include an example of use? You have included all the necessary components, but how would one use this in practice? A demonstration in a `main` function would be ideal.

ash-teach commented on patterns/observers/Observer.kt:
```diff
@@ -0,0 +1,76 @@
+  // creates a list of observers
+  private val observersList = ArrayList<Observers>()
+  // creates a default state
+  private var state: String = ""
```
It would be more idiomatic to handle this with a nullable: https://kotlinlang.org/docs/basic-syntax.html#nullable-values-and-null-checks

ash-teach commented on patterns/observers/Observer.kt:
```diff
@@ -0,0 +1,76 @@
+  override fun getState(): String {
+      return this.state
+  }
+}
```
It may be more appropriate to use getter and setter properties for this: https://kotlinlang.org/docs/properties.html#getters-and-setters

ash-teach commented:
I've added a couple comments to consider in review. I don't think either are 100% necessary for demonstrating this particular Observer implementation, but they may be more Kotlin idiomatic. 

I've added @abhiram-shaji as an assignee to this task because he's take DGL 204 and he should have some insight into typical Kotlin idioms.

ash-teach commented on patterns/observers/Observer.kt:
```diff
@@ -0,0 +1,76 @@
+
+class Subject : Subjects {
+  // creates a list of observers
+  private val observersList = ArrayList<Observers>()
```
`ArrayList` or `MutableList`? `MutableList` might be more appropriate, given the generality of this example. Here's a thread that describes the difference succinctly: https://www.reddit.com/r/Kotlin/comments/kv15dx/difference_between_a_mutablelist_and_an_arraylist/

Evelynn8Bit committed:
Change ArrayList -> MutableList && state: String -> state: String?

Dhuds1 committed:
Merge branch 'nic-dgl104-winter-2024:main' into dhuds-dev

Dhuds1 commented:
@ash-teach I have added documentation stating where I got help from. I also changed the `ArrayList` to `MutableList` as only the contents of the list is important. I also made the `subject state nullable` "I think?" And finally, I added a `main.kt` file to showcase a basic use case.

I did attempted to create a `get()` and `set()` on the `state` but I was getting IDE errors saying `redundant` or compiler `JVM errors`, so if anyone one else can improve this code please go for it! I believe this is the extent of my Kotlin abilities!
